File: app/src/main/java/tpcreative/co/qrscanner/common/api/RootAPI.java
package tpcreative.co.qrscanner.common.api;
import java.util.Map;
import io.reactivex.Observable;
import okhttp3.MultipartBody;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.http.DELETE;
import retrofit2.http.FieldMap;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Headers;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.Part;
import retrofit2.http.Path;
import retrofit2.http.Query;
import retrofit2.http.Streaming;
import tpcreative.co.qrscanner.common.api.response.BaseResponse;
import tpcreative.co.qrscanner.common.api.response.DriveResponse;
import tpcreative.co.qrscanner.model.DriveAbout;

public interface RootAPI{
    String ROOT_GOOGLE_DRIVE = "https://www.googleapis.com/";
    String CHECKOUT = "/api/checkout/syncDevices";
    String AUTHOR = "/api/author/syncDevices";
    String CHECK_VERSION = "/api/author/version";
    String GET_DRIVE_ABOUT = "/drive/v3/about?fields=user,storageQuota,kind";
    String UPLOAD_FILE_TO_GOOGLE_DRIVE = "/upload/drive/v3/files?uploadType=multipart";
    String DOWNLOAD_FILE_FROM_GOOGLE_DRIVE = "/drive/v3/files/{id}?alt=media";
    String GET_LIST_FILE_IN_APP_FOLDER = "/drive/v3/files";
    String GET_FILES_INFO = "/drive/v3/files/{id}";
    String DELETE_CLOUD_ITEM = "/drive/v3/files/{id}";
    @FormUrlEncoded
    @POST(CHECKOUT)
    Observable<BaseResponse> onCheckout(@FieldMap Map<String, String> request);


    @FormUrlEncoded
    @POST(AUTHOR)
    Observable<BaseResponse> onAuthor(@FieldMap Map<String, String> request);


    @POST(CHECK_VERSION)
    Observable<BaseResponse> onCheckVersion();

    @Headers({"Accept: application/json"})
    @GET(GET_DRIVE_ABOUT)
    Observable<DriveAbout> onGetDriveAbout(@Header("Authorization") String token);

    @POST(UPLOAD_FILE_TO_GOOGLE_DRIVE)
    @Multipart
    Call<DriveResponse> uploadFileMultipleInAppFolder(
            @Header("Authorization") String authToken,
            @Part MultipartBody.Part metaPart,
            @Part MultipartBody.Part dataPart,
            @Query("type") String type);

    @GET(DOWNLOAD_FILE_FROM_GOOGLE_DRIVE)
    @Streaming
    Observable<Response<ResponseBody>> downloadDriveFile(@Header("Authorization") String authToken, @Path("id") String id);

    @Headers({"Accept: application/json"})
    @GET(GET_LIST_FILE_IN_APP_FOLDER)
    Observable<DriveAbout> onGetListFileInAppFolder(@Header("Authorization") String token,@Query("spaces")String value);

    @Headers({"Accept: application/json"})
    @DELETE(DELETE_CLOUD_ITEM)
    Observable<Response<DriveAbout>> onDeleteCloudItem(@Header("Authorization") String token, @Path("id") String id);
}

File: app/src/main/java/tpcreative/co/qrscanner/common/api/RootAPI.kt
package tpcreative.co.qrscanner.common.api
import okhttp3.MultipartBody
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Response
import retrofit2.http.*
import tpcreative.co.qrscanner.common.api.request.CheckoutRequest
import tpcreative.co.qrscanner.common.api.response.BaseResponse
import tpcreative.co.qrscanner.common.api.response.DriveResponse
import tpcreative.co.qrscanner.common.api.response.RootResponse
import tpcreative.co.qrscanner.model.DriveAbout

interface RootAPI {
    @POST(CHECKOUT)
    suspend fun onCheckout(@Body request: CheckoutRequest?): RootResponse?
    @FormUrlEncoded
    @POST(AUTHOR)
    suspend fun onAuthor(@FieldMap request: MutableMap<String?, String?>?): BaseResponse?
    @GET
    suspend fun onCheckVersion(@Url url: String?): BaseResponse?
    @Headers("Accept: application/json")
    @GET(GET_DRIVE_ABOUT)
    suspend fun onGetDriveAbout(@Header("Authorization") token: String?): DriveAbout?
    @POST(UPLOAD_FILE_TO_GOOGLE_DRIVE)
    @Multipart
    suspend fun uploadFileMultipleInAppFolder(
            @Header("Authorization") authToken: String?,
            @Part metaPart: MultipartBody.Part?,
            @Part dataPart: MultipartBody.Part?,
            @Query("type") type: String?): Call<DriveResponse?>?

    @GET(DOWNLOAD_FILE_FROM_GOOGLE_DRIVE)
    @Streaming
    suspend fun downloadDriveFile(@Header("Authorization") authToken: String?, @Path("id") id: String?): Response<ResponseBody?>
    @Headers("Accept: application/json")
    @GET(GET_LIST_FILE_IN_APP_FOLDER)
    suspend fun onGetListFileInAppFolder(@Header("Authorization") token: String?, @Query("spaces") value: String?): DriveAbout?
    @Headers("Accept: application/json")
    @DELETE(DELETE_CLOUD_ITEM)
    suspend fun onDeleteCloudItem(@Header("Authorization") token: String?, @Path("id") id: String?): Response<DriveAbout>?

    companion object {
        const val ROOT_GOOGLE_DRIVE: String = "https://www.googleapis.com/"
        const val CHECKOUT: String = "/api/qrscanner/checkout/transaction"
        const val AUTHOR: String = "/api/author/syncDevices"
        const val CHECK_VERSION: String = "http://tpcreative.me/qrscanner.php"
        const val GET_DRIVE_ABOUT: String = "/drive/v3/about?fields=user,storageQuota,kind"
        const val UPLOAD_FILE_TO_GOOGLE_DRIVE: String = "/upload/drive/v3/files?uploadType=multipart"
        const val DOWNLOAD_FILE_FROM_GOOGLE_DRIVE: String = "/drive/v3/files/{id}?alt=media"
        const val GET_LIST_FILE_IN_APP_FOLDER: String = "/drive/v3/files"
        const val GET_FILES_INFO: String = "/drive/v3/files/{id}"
        const val DELETE_CLOUD_ITEM: String = "/drive/v3/files/{id}"
    }
}